iotfunctions/bif.py

```diff
@@ -2816,6 +2816,166 @@ def execute(self, df):
+    Watson Machine Learning. The results are passed back to the univariate
+    output_items column.
+    Credentials for the WML endpoint representing the deployed function are stored
+    as pipeline constants, a name to lookup the WML credentials as JSON document.
+    Example: 'my_deployed_endpoint_wml_credentials' referring to
+    {
+	    "apikey": "<my api key",
+	    "url": "https://us-south.ml.cloud.ibm.com",
+	    "space_id": "<my space id>",
+	    "deployment_id": "<my deployment id">
+    }
+    This name is passed to InvokeWMLModel in wml_auth.
+    '''
+    def __init__(self, input_items, wml_auth, output_items):
+        super().__init__()
+
+        logger.debug(input_items)
+
+        self.whoami = 'InvokeWMLModel'
+
+        self.input_items = input_items
+        self.output_items = output_items
+        self.wml_auth = wml_auth
+
+        self.deployment_id = None
+        self.apikey = None
+        self.wml_endpoint = None
+        self.space_id = None
+
+        self.client = None
+
+        self.logged_on = False
+
+
+    def __str__(self):
+        out = self.__class__.__name__
+        try:
+            out = out + 'Input: ' + str(self.input_items) + '\n'
+            out = out + 'Output: ' + str(self.output_items) + '\n'
+
+            if self.wml_auth is not None:
+                out = out + 'WML auth: ' + str(self.wml_auth) + '\n'
+            else:
+                out = out + 'APIKey: ' + str(self.apikey) + '\n'
+                out = out + 'WML endpoint: ' + str(self.wml_endpoint) + '\n'
+                out = out + 'WML space id: ' + str(self.space_id) + '\n'
+                out = out + 'WML deployment id: ' + str(self.deployment_id) + '\n'
+        except Exception:
+            pass
+        return out
+
+
+    def login(self):
+
+        # only do it once
+        if self.logged_on:
+            return
+
+        # retrieve WML credentials as constant
+        #    {"apikey": api_key, "url": 'https://' + location + '.ml.cloud.ibm.com'}
+        if self.wml_auth is not None:
+            c = self._entity_type.get_attributes_dict()
+            try:
+                wml_credentials = c[self.wml_auth]
+                print('WML Credentials ' , str(wml_credentials))
+            except Exception as ae:
+                wml_credentials = {'apikey': self.apikey , 'url': self.wml_endpoint, 'space_id': self.space_id}
+                logger.error('WML Credentials constant ' + self.wml_auth + ' not present. Error ' + str(ae))
+                pass
+            self.deployment_id = wml_credentials['deployment_id']
+            self.space_id = wml_credentials['space_id']
+        else:
+            wml_credentials = {'apikey': self.apikey , 'url': self.wml_endpoint, 'space_id': self.space_id}
+
+        # get client and check credentials
+        self.client = APIClient(wml_credentials)
+        if self.client is None:
+            logger.error('WML API Key invalid')
```

Comment:
logger.error() in conjunction with 'raise Exception' leaves the stacktrace twice in logfile what is umfavourable. I would omit the logger.error() statement.